docs/excel/custom-functions-batching.md

```diff
@@ -0,0 +1,214 @@
+
+```typescript
+interface IBatchEntry {
+  operation: string;
+  args: any[];
+  resolve: (data: any) => void;
+  reject: (error: Error) => void;
+}
+```
+
+Next we create the batch array based on the previous interface. Also we create an `_isBatchedRequestSchedule variable to track if a batch is scheduled or not. This will be important later for timing batch calls to the remote service.
+
+```typescript
+const _batch: IBatchEntry[] = [];
+let _isBatchedRequestScheduled = false;
+```
+
+Finally when Excel calls your custom function, you need to push the operation into the batch array. The following code shows how to add a new operation from a custom function. It creates a new batch entry, creates a new promise to resolve or reject the operation, and pushes the entry into the batch array.
+
+This code also checks to see if a batch is scheduled. In this case each batch is scheduled to run every 100ms. You can adjust this value as needed. Higher values will result in bigger batches being sent to the remote service, and a longer wait time for the user to see results. Lower values will tend to send more batches to the remote service, but with a quick response time for users.
+
+### Add the _pushOperation function to functions.ts
+
+```typescript
+function _pushOperation(op: string, args: any[]) {
+  // Create an entry for your custom function.
+  const invocationEntry: IBatchEntry = {
+    operation: op, // e.g. sum
+    args: args,
+    resolve: undefined,
+    reject: undefined,
+  };
+
+  // Create a unique promise for this invocation,
+  // and save its resolve and reject functions into the invocation entry.
+  const promise = new Promise((resolve, reject) => {
+    invocationEntry.resolve = resolve;
+    invocationEntry.reject = reject;
+  });
+
+  // Push the invocation entry into the next batch.
+  _batch.push(invocationEntry);
+
+  // If a remote request hasn't been scheduled yet,
+  // schedule it after a certain timeout, e.g. 100 ms.
+  if (!_isBatchedRequestScheduled) {
+    _isBatchedRequestScheduled = true;
+    setTimeout(_makeRemoteRequest, 100);
+  }
+
+  // Return the promise for this invocation.
+  return promise;
+}
+```
+
+## Make the remote request
+
+The purpose of the `_makeRemoteRequest` function is to pass the batch of operations to the remote service, and then return the results back to each custom function. It first creates a copy of the batch array. This allows concurrent custom function calls from Excel to immediately begin batching in a new array. The copy is then turned into a simpler array that does not contain the promise information. It wouldn't make sense to pass the promises to a remote service since they would not work. The `_makeRemoteRequest will either reject or resolve each promise based on what the remote service returns.
```

Comment:
...the remote service, and then return the results to each custom function. It first...